src/access/AccessRestriction.cpp

```diff
@@ -0,0 +1,280 @@
+{
+    ChipLogProgress(DataManagement, "AccessRestriction: action %d", static_cast<int>(requestPath.requestType));
+
+    if (requestPath.requestType == RequestType::kRequestTypeUnknown)
+    {
+        ChipLogError(DataManagement, "AccessRestriction: RequestPath type is unknown");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    // wildcard event subscriptions are allowed since wildcard is only used when setting up the subscription and
+    // we want that request to succeed (when generating the report, this method will be called with the specific
+    // event id). All other requests require an entity id
+    if (!requestPath.entityId.has_value())
+    {
+        if (requestPath.requestType == RequestType::kEventReadOrSubscribeRequest)
+        {
+            return CHIP_NO_ERROR;
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    for (auto & entry : mFabricEntries[subjectDescriptor.fabricIndex])
+    {
+        if (entry->endpointNumber != requestPath.endpoint || entry->clusterId != requestPath.cluster)
+        {
+            continue;
+        }
+
+        for (auto & restriction : entry->restrictions)
+        {
+            // a missing id is a wildcard
+            bool idMatch = !restriction.id.HasValue() || restriction.id.Value() == requestPath.entityId.value();
+            if (!idMatch)
+            {
+                continue;
+            }
+
+            switch (restriction.restrictionType)
+            {
+            case Type::kAttributeAccessForbidden:
+                if (requestPath.requestType == RequestType::kAttributeReadRequest ||
+                    requestPath.requestType == RequestType::kAttributeWriteRequest)
+                {
+                    return CHIP_ERROR_ACCESS_DENIED;
+                }
+                break;
+            case Type::kAttributeWriteForbidden:
+                if (requestPath.requestType == RequestType::kAttributeWriteRequest)
+                {
+                    return CHIP_ERROR_ACCESS_DENIED;
+                }
+                break;
+            case Type::kCommandForbidden:
+                if (requestPath.requestType == RequestType::kCommandInvokeRequest)
+                {
+                    return CHIP_ERROR_ACCESS_DENIED;
+                }
+                break;
+            case Type::kEventForbidden:
+                if (requestPath.requestType == RequestType::kEventReadOrSubscribeRequest)
+                {
+                    return CHIP_ERROR_ACCESS_DENIED;
+                }
+                break;
+            }
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void AccessRestriction::ClearData()
+{
+    // iterate over mFabricEntries and call DeleteEntry for each entry
+    for (auto & fabricEntry : mFabricEntries)
+    {
+        for (size_t i = 0; i < fabricEntry.second.size(); i++)
+        {
+            DeleteEntry(i, fabricEntry.first);
+        }
+    }
```

Comment:
This needs to be done in reverse order otherwise you will immediately run through the end.

Reply:
This is no longer relevant given the data model management change.